File: sdks/python/src/bcs/deserializer.py
"""BCS Deserializer - Manual deserialization API."""

from __future__ import annotations

import struct
from typing import Any, Callable, TypeVar

from .errors import (
    ExceededContainerDepth,
    ExceededMaxLength,
    InvalidBoolean,
    InvalidOption,
    InvalidUtf8,
    NonCanonicalMap,
    NonCanonicalUleb128,
    RemainingInput,
    Uleb128Overflow,
    UnexpectedEof,
)
from .types import MAX_CONTAINER_DEPTH, MAX_SEQUENCE_LENGTH

T = TypeVar("T")

# Pre-compiled struct formats for performance
_STRUCT_U16 = struct.Struct("<H")
_STRUCT_U32 = struct.Struct("<I")
_STRUCT_U64 = struct.Struct("<Q")
_STRUCT_I8 = struct.Struct("<b")
_STRUCT_I16 = struct.Struct("<h")
_STRUCT_I32 = struct.Struct("<i")
_STRUCT_I64 = struct.Struct("<q")


class BcsDeserializer:
    """Manual BCS deserialization API.

    Provides explicit methods for deserializing each BCS type.
    Use this for full control over deserialization.

    Example:
        >>> d = BcsDeserializer(data)
        >>> value1 = d.read_u8()
        >>> value2 = d.read_u64()
        >>> value3 = d.read_string()
        >>> d.check_end()  # Verify no remaining bytes
    """

    __slots__ = ("_data", "_offset", "_len", "_max_depth", "_current_depth", "_max_alloc")

    def __init__(
        self,
        data: bytes | bytearray | memoryview,
        max_depth: int = MAX_CONTAINER_DEPTH,
        max_alloc: int | None = None,
    ) -> None:
        """Initialize a new deserializer.

        Args:
            data: Bytes to deserialize from
            max_depth: Maximum container nesting depth (default: 500)
            max_alloc: Maximum allocation size for vectors/bytes (default: MAX_SEQUENCE_LENGTH).
                       Set to a smaller value for defense-in-depth against memory exhaustion.
        """
        self._data = memoryview(data) if not isinstance(data, memoryview) else data
        self._offset = 0
        self._len = len(data)
        self._max_depth = max_depth
        self._current_depth = 0
        self._max_alloc = max_alloc if max_alloc is not None else MAX_SEQUENCE_LENGTH

    @property
    def remaining(self) -> int:
        """Number of bytes remaining to read."""
        return self._len - self._offset

    def check_end(self) -> None:
        """Verify all input has been consumed.

        Raises:
            RemainingInput: If bytes remain after deserialization
        """
        remaining = self._len - self._offset
        if remaining > 0:
            raise RemainingInput(remaining)

    def _check_depth(self, container: str = "") -> None:
        """Check and increment container depth."""
        if self._current_depth >= self._max_depth:
            raise ExceededContainerDepth(container)
        self._current_depth += 1

    def _leave_container(self) -> None:
        """Decrement container depth."""
        if self._current_depth > 0:
            self._current_depth -= 1

    # =========================================================================
    # BOOLEAN
    # =========================================================================

    def read_bool(self) -> bool:
        """Deserialize a boolean.

        Returns:
            Boolean value

        Raises:
            UnexpectedEof: If no bytes available
            InvalidBoolean: If byte is not 0x00 or 0x01
        """
        if self._offset >= self._len:
            raise UnexpectedEof(1, 0)
        byte = self._data[self._offset]
        self._offset += 1
        if byte == 0:
            return False
        elif byte == 1:
            return True
        else:
            raise InvalidBoolean(byte)

    # =========================================================================
    # UNSIGNED INTEGERS
    # =========================================================================

    def read_u8(self) -> int:
        """Deserialize an unsigned 8-bit integer.

        Returns:
            Integer in range [0, 255]
        """
        if self._offset >= self._len:
            raise UnexpectedEof(1, 0)
        value = self._data[self._offset]
        self._offset += 1
        return value

    def read_u16(self) -> int:
        """Deserialize an unsigned 16-bit integer (little-endian).

        Returns:
            Integer in range [0, 65535]
        """
        end = self._offset + 2
        if end > self._len:
            raise UnexpectedEof(2, self._len - self._offset)
        value = _STRUCT_U16.unpack(self._data[self._offset : end])[0]
        self._offset = end
        return value

    def read_u32(self) -> int:
        """Deserialize an unsigned 32-bit integer (little-endian).

        Returns:
            Integer in range [0, 2^32-1]
        """
        end = self._offset + 4
        if end > self._len:
            raise UnexpectedEof(4, self._len - self._offset)
        value = _STRUCT_U32.unpack(self._data[self._offset : end])[0]
        self._offset = end
        return value

    def read_u64(self) -> int:
        """Deserialize an unsigned 64-bit integer (little-endian).

        Returns:
            Integer in range [0, 2^64-1]
        """
        end = self._offset + 8
        if end > self._len:
            raise UnexpectedEof(8, self._len - self._offset)
        value = _STRUCT_U64.unpack(self._data[self._offset : end])[0]
        self._offset = end
        return value

    def read_u128(self) -> int:
        """Deserialize an unsigned 128-bit integer (little-endian).

        Returns:
            Integer in range [0, 2^128-1]
        """
        end = self._offset + 16
        if end > self._len:
            raise UnexpectedEof(16, self._len - self._offset)
        value = int.from_bytes(self._data[self._offset : end], "little")
        self._offset = end
        return value

    def read_u256(self) -> int:
        """Deserialize an unsigned 256-bit integer (little-endian).

        Returns:
            Integer in range [0, 2^256-1]
        """
        end = self._offset + 32
        if end > self._len:
            raise UnexpectedEof(32, self._len - self._offset)
        value = int.from_bytes(self._data[self._offset : end], "little")
        self._offset = end
        return value

    # =========================================================================
    # SIGNED INTEGERS (two's complement)
    # =========================================================================

    def read_i8(self) -> int:
        """Deserialize a signed 8-bit integer (two's complement).

        Returns:
            Integer in range [-128, 127]
        """
        if self._offset >= self._len:
            raise UnexpectedEof(1, 0)
        value = _STRUCT_I8.unpack(self._data[self._offset : self._offset + 1])[0]
        self._offset += 1
        return value

    def read_i16(self) -> int:
        """Deserialize a signed 16-bit integer (two's complement, little-endian).

        Returns:
            Integer in range [-32768, 32767]
        """
        end = self._offset + 2
        if end > self._len:
            raise UnexpectedEof(2, self._len - self._offset)
        value = _STRUCT_I16.unpack(self._data[self._offset : end])[0]
        self._offset = end
        return value

    def read_i32(self) -> int:
        """Deserialize a signed 32-bit integer (two's complement, little-endian).

        Returns:
            Integer in range [-2^31, 2^31-1]
        """
        end = self._offset + 4
        if end > self._len:
            raise UnexpectedEof(4, self._len - self._offset)
        value = _STRUCT_I32.unpack(self._data[self._offset : end])[0]
        self._offset = end
        return value

    def read_i64(self) -> int:
        """Deserialize a signed 64-bit integer (two's complement, little-endian).

        Returns:
            Integer in range [-2^63, 2^63-1]
        """
        end = self._offset + 8
        if end > self._len:
            raise UnexpectedEof(8, self._len - self._offset)
        value = _STRUCT_I64.unpack(self._data[self._offset : end])[0]
        self._offset = end
        return value

    def read_i128(self) -> int:
        """Deserialize a signed 128-bit integer (two's complement, little-endian).

        Returns:
            Integer in range [-2^127, 2^127-1]
        """
        end = self._offset + 16
        if end > self._len:
            raise UnexpectedEof(16, self._len - self._offset)
        value = int.from_bytes(self._data[self._offset : end], "little", signed=True)
        self._offset = end
        return value

    def read_i256(self) -> int:
        """Deserialize a signed 256-bit integer (two's complement, little-endian).

        Returns:
            Integer in range [-2^255, 2^255-1]
        """
        end = self._offset + 32
        if end > self._len:
            raise UnexpectedEof(32, self._len - self._offset)
        value = int.from_bytes(self._data[self._offset : end], "little", signed=True)
        self._offset = end
        return value

    # =========================================================================
    # ULEB128
    # =========================================================================

    def read_uleb128(self) -> int:
        """Deserialize a ULEB128-encoded unsigned integer.

        Returns:
            Non-negative integer fitting in u32

        Raises:
            UnexpectedEof: If input ends before ULEB128 is complete
            NonCanonicalUleb128: If encoding is not minimal
            Uleb128Overflow: If value exceeds u32 max
        """
        # Inline ULEB128 decoding for performance
        value = 0
        shift = 0
        data = self._data
        offset = self._offset
        data_len = self._len

        for i in range(5):  # Maximum 5 bytes for u32
            if offset + i >= data_len:
                raise UnexpectedEof()

            byte = data[offset + i]
            digit = byte & 0x7F
            value |= digit << shift

            # Check if this is the last byte (high bit is 0)
            if digit == byte:
                # Reject non-canonical encodings (trailing zero bytes)
                if shift > 0 and digit == 0:
                    raise NonCanonicalUleb128()

                # Check for overflow
                if value > 0xFFFFFFFF:
                    raise Uleb128Overflow()

                self._offset = offset + i + 1
                return value

            shift += 7

        # If we get here, we read 5 bytes and all had continuation bits
        raise Uleb128Overflow()

    # =========================================================================
    # BYTES AND STRINGS
    # =========================================================================

    def read_bytes(self) -> bytes:
        """Deserialize a byte array (length-prefixed with ULEB128).

        Returns:
            Deserialized bytes

        Raises:
            ExceededMaxLength: If length exceeds max_alloc or MAX_SEQUENCE_LENGTH
        """
        length = self.read_uleb128()
        if length > self._max_alloc:
            raise ExceededMaxLength(length)
        end = self._offset + length
        if end > self._len:
            raise UnexpectedEof(length, self._len - self._offset)
        result = bytes(self._data[self._offset : end])
        self._offset = end
        return result

    def read_string(self) -> str:
        """Deserialize a UTF-8 string (length-prefixed with ULEB128).

        Returns:
            Deserialized string

        Raises:
            ExceededMaxLength: If length exceeds max_alloc
            InvalidUtf8: If bytes are not valid UTF-8
        """
        length = self.read_uleb128()
        if length > self._max_alloc:
            raise ExceededMaxLength(length)
        end = self._offset + length
        if end > self._len:
            raise UnexpectedEof(length, self._len - self._offset)
        try:
            # Decode directly from memoryview
            result = self._data[self._offset : end].tobytes().decode("utf-8")
        except UnicodeDecodeError as e:
            raise InvalidUtf8(str(e)) from e
        self._offset = end
        return result

    def read_fixed_bytes(self, length: int) -> bytes:
        """Deserialize fixed-length bytes (no length prefix).

        Args:
            length: Number of bytes to read

        Returns:
            Deserialized bytes
        """
        end = self._offset + length
        if end > self._len:
            raise UnexpectedEof(length, self._len - self._offset)
        result = bytes(self._data[self._offset : end])
        self._offset = end
        return result

    # =========================================================================
    # OPTION
    # =========================================================================

    def read_option(self, deserializer: Callable[["BcsDeserializer"], T]) -> T | None:
        """Deserialize an optional value.

        Args:
            deserializer: Function to deserialize the value if present

        Returns:
            The value or None

        Raises:
            InvalidOption: If tag byte is not 0x00 or 0x01
        """
        if self._offset >= self._len:
            raise UnexpectedEof(1, 0)
        tag = self._data[self._offset]
        self._offset += 1
        if tag == 0:
            return None
        elif tag == 1:
            return deserializer(self)
        else:
            raise InvalidOption(tag)

    def read_option_u8(self) -> int | None:
        """Deserialize an optional u8 (optimized)."""
        if self._offset >= self._len:
            raise UnexpectedEof(1, 0)
        tag = self._data[self._offset]
        self._offset += 1
        if tag == 0:
            return None
        elif tag == 1:
            return self.read_u8()
        else:
            raise InvalidOption(tag)

    def read_option_u64(self) -> int | None:
        """Deserialize an optional u64 (optimized)."""
        if self._offset >= self._len:
            raise UnexpectedEof(1, 0)
        tag = self._data[self._offset]
        self._offset += 1
        if tag == 0:
            return None
        elif tag == 1:
            return self.read_u64()
        else:
            raise InvalidOption(tag)

    def read_option_string(self) -> str | None:
        """Deserialize an optional string."""
        return self.read_option(lambda d: d.read_string())

    # =========================================================================
    # VECTOR
    # =========================================================================

    def read_vector(self, deserializer: Callable[["BcsDeserializer"], T]) -> list[T]:
        """Deserialize a vector of values.

        Args:
            deserializer: Function to deserialize each element

        Returns:
            List of deserialized values

        Raises:
            ExceededMaxLength: If length exceeds max_alloc
        """
        length = self.read_uleb128()
        if length > self._max_alloc:
            raise ExceededMaxLength(length)
        return [deserializer(self) for _ in range(length)]

    def read_vector_u8(self) -> list[int]:
        """Deserialize a vector of u8 (optimized - returns list of ints)."""
        length = self.read_uleb128()
        if length > self._max_alloc:
            raise ExceededMaxLength(length)
        end = self._offset + length
        if end > self._len:
            raise UnexpectedEof(length, self._len - self._offset)
        result = list(self._data[self._offset : end])
        self._offset = end
        return result

    def read_vector_u64(self) -> list[int]:
        """Deserialize a vector of u64 (optimized)."""
        length = self.read_uleb128()
        if length > self._max_alloc:
            raise ExceededMaxLength(length)
        # Check available bytes before computing byte_length to prevent overflow
        available = self._len - self._offset
        if length > available // 8:
            raise UnexpectedEof(length * 8, available)
        byte_length = length * 8
        end = self._offset + byte_length

        # Use iter_unpack for better performance on large vectors
        data = self._data[self._offset : end]
        result = [v for (v,) in struct.iter_unpack("<Q", data)]
        self._offset = end
        return result

    def read_vector_string(self) -> list[str]:
        """Deserialize a vector of strings."""
        return self.read_vector(lambda d: d.read_string())

    # =========================================================================
    # ENUM
    # =========================================================================

    def read_variant_index(self) -> int:
        """Read an enum variant index (ULEB128).

        Returns:
            Variant index (0-based)
        """
        self._check_depth("enum")
        return self.read_uleb128()

    def leave_enum(self) -> None:
        """Signal end of enum deserialization (for depth tracking)."""
        self._leave_container()

    # =========================================================================
    # STRUCT
    # =========================================================================

    def enter_struct(self, name: str = "") -> None:
        """Signal start of struct deserialization (for depth tracking).

        Args:
            name: Optional struct name for error messages
        """
        self._check_depth(name)

    def leave_struct(self) -> None:
        """Signal end of struct deserialization (for depth tracking)."""
        self._leave_container()

    # =========================================================================
    # MAP
    # =========================================================================

    def read_map(
        self,
        key_deserializer: Callable[["BcsDeserializer"], Any],
        value_deserializer: Callable[["BcsDeserializer"], Any],
    ) -> dict[Any, Any]:
        """Deserialize a map (verifying sorted keys).

        Args:
            key_deserializer: Function to deserialize keys
            value_deserializer: Function to deserialize values

        Returns:
            Deserialized dictionary

        Raises:
            ExceededMaxLength: If length exceeds max_alloc
            NonCanonicalMap: If keys are not sorted or contain duplicates
        """
        length = self.read_uleb128()
        if length > self._max_alloc:
            raise ExceededMaxLength(length)

        result: dict[Any, Any] = {}
        prev_key_bytes: bytes | None = None

        for _ in range(length):
            # Record position before reading key
            start_offset = self._offset

            key = key_deserializer(self)

            # Get the key bytes for comparison
            key_bytes = bytes(self._data[start_offset : self._offset])

            # Verify sorted order and no duplicates
            if prev_key_bytes is not None:
                if key_bytes <= prev_key_bytes:
                    if key_bytes == prev_key_bytes:
                        raise NonCanonicalMap("duplicate key")
                    else:
                        raise NonCanonicalMap("keys not sorted")

            prev_key_bytes = key_bytes
            value = value_deserializer(self)
            result[key] = value

        return result

File: sdks/ruby/lib/bcs/deserializer.rb
# frozen_string_literal: true

module BCS
  # BCS Deserializer - Manual deserialization API
  # Optimized to use String with binary encoding and native unpack operations
  class Deserializer

    # Pre-computed sign bit and modulus constants for signed integers
    SIGN_BIT_128 = 1 << 127
    MODULUS_128 = 1 << 128
    SIGN_BIT_256 = 1 << 255
    MODULUS_256 = 1 << 256

    def initialize(data)
      @data = data.is_a?(String) ? data.b : data.pack("C*")
      @offset = 0
      @depth = 0
    end

    # ========================================================================
    # BOOLEAN
    # ========================================================================

    # Read a boolean value
    # @return [Boolean]
    def read_bool
      byte = read_u8
      case byte
      when 0 then false
      when 1 then true
      else raise Error.invalid_boolean(byte)
      end
    end

    # ========================================================================
    # UNSIGNED INTEGERS
    # ========================================================================

    # Read an unsigned 8-bit integer
    def read_u8
      check_remaining(1)
      value = @data.getbyte(@offset)
      @offset += 1
      value
    end

    # Read an unsigned 16-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_u16
      check_remaining(2)
      value = @data.unpack1("@#{@offset}v")
      @offset += 2
      value
    end

    # Read an unsigned 32-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_u32
      check_remaining(4)
      value = @data.unpack1("@#{@offset}V")
      @offset += 4
      value
    end

    # Read an unsigned 64-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_u64
      check_remaining(8)
      value = @data.unpack1("@#{@offset}Q<")
      @offset += 8
      value
    end

    # Read an unsigned 128-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_u128
      check_remaining(16)
      low, high = @data.unpack("@#{@offset}Q<Q<")
      @offset += 16
      low | (high << 64)
    end

    # Read an unsigned 256-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_u256
      check_remaining(32)
      parts = @data.unpack("@#{@offset}Q<Q<Q<Q<")
      @offset += 32
      parts[0] | (parts[1] << 64) | (parts[2] << 128) | (parts[3] << 192)
    end

    # ========================================================================
    # SIGNED INTEGERS
    # ========================================================================

    # Read a signed 8-bit integer
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_i8
      check_remaining(1)
      value = @data.unpack1("@#{@offset}c")
      @offset += 1
      value
    end

    # Read a signed 16-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_i16
      check_remaining(2)
      value = @data.unpack1("@#{@offset}s<")
      @offset += 2
      value
    end

    # Read a signed 32-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_i32
      check_remaining(4)
      value = @data.unpack1("@#{@offset}l<")
      @offset += 4
      value
    end

    # Read a signed 64-bit integer (little-endian)
    # Optimized using unpack with offset specifier to avoid string slice allocation
    def read_i64
      check_remaining(8)
      value = @data.unpack1("@#{@offset}q<")
      @offset += 8
      value
    end

    # Read a signed 128-bit integer (little-endian)
    # Uses pre-computed constants for better performance
    def read_i128
      value = read_u128
      value >= SIGN_BIT_128 ? value - MODULUS_128 : value
    end

    # Read a signed 256-bit integer (little-endian)
    # Uses pre-computed constants for better performance
    def read_i256
      value = read_u256
      value >= SIGN_BIT_256 ? value - MODULUS_256 : value
    end

    # ========================================================================
    # ULEB128
    # ========================================================================

    # Read a ULEB128-encoded value
    # Optimized with inline fast path for single-byte values
    def read_uleb128
      check_remaining(1)
      byte = @data.getbyte(@offset)

      # Fast path: single byte (values 0-127)
      if byte < 0x80
        @offset += 1
        return byte
      end

      # Slow path: multi-byte value
      value, bytes_read = Uleb128.decode(@data, @offset)
      @offset += bytes_read
      value
    end

    # ========================================================================
    # BYTES AND STRINGS
    # ========================================================================

    # Read fixed-length bytes (without length prefix)
    # @param length [Integer] Number of bytes to read
    # @return [Array<Integer>] The bytes as an array
    def read_fixed_bytes(length)
      check_remaining(length)
      result = @data.byteslice(@offset, length).bytes
      @offset += length
      result
    end

    # Read fixed-length bytes as a binary string (without length prefix)
    # @param length [Integer] Number of bytes to read
    # @return [String] The bytes as a binary string
    def read_fixed_bytes_as_string(length)
      check_remaining(length)
      result = @data.byteslice(@offset, length)
      @offset += length
      result
    end

    # Read bytes with ULEB128 length prefix
    # @return [Array<Integer>] The bytes
    def read_bytes
      length = read_uleb128
      check_sequence_length(length)
      read_fixed_bytes(length)
    end

    # Read bytes with ULEB128 length prefix as a binary string
    # @return [String] The bytes as a binary string
    def read_bytes_as_string
      length = read_uleb128
      check_sequence_length(length)
      read_fixed_bytes_as_string(length)
    end

    # Read a UTF-8 string with ULEB128 length prefix
    # @return [String] The string
    def read_string
      str = read_bytes_as_string.force_encoding("UTF-8")
      raise Error.invalid_utf8 unless str.valid_encoding?

      str
    rescue Encoding::InvalidByteSequenceError
      raise Error.invalid_utf8
    end

    # ========================================================================
    # COMPOSITE TYPES
    # ========================================================================

    # Read an optional value
    # @yield [Deserializer] Block to deserialize the value if present
    # @return [Object, nil] The deserialized value or nil
    def read_option(&deserializer)
      tag = read_u8
      case tag
      when 0 then nil
      when 1 then deserializer.call(self)
      else raise Error.invalid_option(tag)
      end
    end

    # Read a vector with element deserializer
    # @yield [Deserializer] Block to deserialize each element
    # @return [Array] The array of deserialized values
    def read_vector(&deserializer)
      length = read_uleb128
      check_sequence_length(length)
      Array.new(length) { deserializer.call(self) }
    end

    # ========================================================================
    # BATCH OPERATIONS (Optimized for common vector types)
    # ========================================================================

    # Read a vector of u8 values (optimized batch operation)
    # @return [Array<Integer>] The array of u8 values
    def read_u8_array
      length = read_uleb128
      check_sequence_length(length)
      read_fixed_bytes(length)
    end

    # Read a vector of u8 values as a binary string (optimized batch operation)
    # @return [String] The bytes as a binary string
    def read_u8_array_as_string
      length = read_uleb128
      check_sequence_length(length)
      read_fixed_bytes_as_string(length)
    end

    # Maximum element counts for batch operations to prevent overflow
    # These ensure byte_length = length * element_size won't overflow
    MAX_U16_ARRAY_LENGTH = MAX_SEQUENCE_LENGTH / 2
    MAX_U32_ARRAY_LENGTH = MAX_SEQUENCE_LENGTH / 4
    MAX_U64_ARRAY_LENGTH = MAX_SEQUENCE_LENGTH / 8

    # Read a vector of u16 values (optimized batch operation)
    # @return [Array<Integer>] The array of u16 values
    def read_u16_array
      length = read_uleb128
      raise Error.exceeded_max_length(length) if length > MAX_U16_ARRAY_LENGTH

      byte_length = length * 2
      check_remaining(byte_length)
      result = @data.unpack("@#{@offset}v#{length}")
      @offset += byte_length
      result
    end

    # Read a vector of u32 values (optimized batch operation)
    # @return [Array<Integer>] The array of u32 values
    def read_u32_array
      length = read_uleb128
      raise Error.exceeded_max_length(length) if length > MAX_U32_ARRAY_LENGTH

      byte_length = length * 4
      check_remaining(byte_length)
      result = @data.unpack("@#{@offset}V#{length}")
      @offset += byte_length
      result
    end

    # Read a vector of u64 values (optimized batch operation)
    # @return [Array<Integer>] The array of u64 values
    def read_u64_array
      length = read_uleb128
      raise Error.exceeded_max_length(length) if length > MAX_U64_ARRAY_LENGTH

      byte_length = length * 8
      check_remaining(byte_length)
      result = @data.unpack("@#{@offset}Q<#{length}")
      @offset += byte_length
      result
    end

    # Read a vector of strings (optimized batch operation)
    # @return [Array<String>] The array of strings
    def read_string_array
      length = read_uleb128
      check_sequence_length(length)
      Array.new(length) { read_string }
    end

    # Read a map with key/value deserializers
    # @yield [key_des, value_des] Blocks to deserialize keys and values
    # @return [Hash] The deserialized map
    def read_map(&block)
      length = read_uleb128
      check_sequence_length(length)

      result = {}
      prev_key_bytes = nil

      length.times do
        # Remember position before reading key
        key_start = @offset

        key = block.call(self, :key)

        # Get key bytes for ordering check (as binary string for efficient comparison)
        key_bytes = @data.byteslice(key_start, @offset - key_start)

        # Check ordering (lexicographic comparison using native string comparison)
        if prev_key_bytes
          cmp = key_bytes <=> prev_key_bytes
          if cmp <= 0
            raise cmp.zero? ? Error.duplicate_map_key : Error.non_canonical_map
          end
        end
        prev_key_bytes = key_bytes

        value = block.call(self, :value)
        result[key] = value
      end

      result
    end

    # Read an enum variant index (ULEB128)
    def read_variant_index
      enter_container
      read_uleb128
    end

    # ========================================================================
    # CONTAINER DEPTH
    # ========================================================================

    # Enter a struct container for depth tracking
    def enter_struct(_name = "")
      enter_container
      self
    end

    # Leave the current struct container
    def leave_struct
      leave_container
      self
    end

    # Enter an enum container and read variant index
    def enter_enum
      read_variant_index
    end

    # Leave the current enum container
    def leave_enum
      leave_container
      self
    end

    # ========================================================================
    # STATE
    # ========================================================================

    # Check that all input has been consumed
    # @raise [Error] if there are remaining bytes
    def check_end
      raise Error.remaining_input(remaining) if @offset < @data.bytesize
    end

    # Get the current offset
    # @return [Integer]
    attr_reader :offset

    # Get the remaining bytes count
    # @return [Integer]
    def remaining
      @data.bytesize - @offset
    end

    # Check if there's more data to read
    # @return [Boolean]
    def remaining?
      @offset < @data.bytesize
    end

    # Get the raw data buffer (for advanced use)
    # @return [String] The raw binary data
    attr_reader :data

    private

    def check_remaining(needed)
      raise ArgumentError, "length must be non-negative" if needed < 0
      raise Error.unexpected_eof if @offset + needed > @data.bytesize
    end

    def check_sequence_length(length)
      raise Error.exceeded_max_length(length) if length > MAX_SEQUENCE_LENGTH
    end

    def enter_container
      @depth += 1
      raise Error.exceeded_container_depth(@depth) if @depth > MAX_CONTAINER_DEPTH
    end

    def leave_container
      @depth -= 1 if @depth.positive?
    end

  end
end
